### PR TITLE
rustdoc: cross-crate re-exports: correctly render late-bound params in source order even if early-bound params are present

### PR DESCRIPTION
**Pull request by fmease**
r? ghost

### CODE REVIEW

**Comment by fmease** [comment]
Let's see how bad perf is for the unoptimized version. However, the benchmark suite might not properly cover this scenario.

@bors try @rust-timer queue

File: src/librustdoc/html/format.rs [hunk 1]
```diff
@@ -188,7 +188,7 @@ impl clean::GenericParamDef {
         cx: &'a Context<'tcx>,
     ) -> impl fmt::Display + 'a + Captures<'tcx> {
         display_fn(move |f| match &self.kind {
-            clean::GenericParamDefKind::Lifetime { outlives } => {
+            clean::GenericParamDefKind::Lifetime { outlives, .. } => {
```
Comment (fmease) [inline src/librustdoc/html/format.rs:191]: revert this change

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -1365,8 +1366,12 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
                 tcx.generics_of(assoc_item.def_id),
                 tcx.explicit_predicates_of(assoc_item.def_id),
             );
-            // FIXME: This does not place parameters in source order (late-bound ones come last)
             generics.params.extend(clean_bound_vars(sig.bound_vars()));
+            // FIXME(fmease): Does this work in the presence of synthetic params (impl-Trait, host effect)?
+            // FIXME(fmease): Optimization: Don't call def_ident_span on non-lifetime params!
+            if !sig.bound_vars().is_empty() {
```
Comment (fmease) [inline src/librustdoc/clean/mod.rs:1372]:
```suggestion
            if !sig.bound_vars().is_empty() && !generics.params.is_empty() {
```

**Comment by fmease** [comment]
> https://github.com/rust-lang/rust/pull/121022#issuecomment-1942026998

Spurious, not a `doc` regression.
@bors rollup=maybe

File: tests/rustdoc/inline_cross/early-late-bound-lifetime-params.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+//
+// For context, at the time of writing early- and late-bound params are stored
+// separately in rustc and therefore rustdoc needs to manually merge them.
+
+#![crate_name = "usr"]
+// aux-crate:dep=early-late-bound-lifetime-params.rs
+// edition:2021
+
+// @has usr/fn.f.html
+// @has - '//pre[@class="rust item-decl"]' "fn f<'a, 'b, 'c, 'd, T, const N: usize>(_: impl Copy)"
```
Comment (fmease) [inline tests/rustdoc/inline_cross/early-late-bound-lifetime-params.rs:12]: Both items used to be rendered like

```rs
pub fn f<'b, 'd, T, const N: usize, 'a, 'c>(_: impl Sized)
``

**Comment by fmease** [comment]
r? rustdoc

**Comment by fmease** [comment]
Eh, since I changed some things around, let's check perf again.
@bors try @rust-timer queue

File: src/librustdoc/clean/types.rs [hunk 4]
```diff
@@ -1339,13 +1339,14 @@ impl GenericParamDefKind {
 
 #[derive(Clone, PartialEq, Eq, Debug, Hash)]
 pub(crate) struct GenericParamDef {
+    pub(crate) did: DefId,
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/types.rs:1342]: Nitpicking: please use `def_id`. We tend to use this over `did` in rustdoc as much as we can and I'd like this trend to continue. :)

File: src/librustdoc/clean/types.rs [hunk 5]
```diff
@@ -1339,13 +1339,14 @@ impl GenericParamDefKind {
...
     pub(crate) name: Symbol,
     pub(crate) kind: GenericParamDefKind,
 }
 
 impl GenericParamDef {
-    pub(crate) fn lifetime(name: Symbol) -> Self {
-        Self { name, kind: GenericParamDefKind::Lifetime { outlives: ThinVec::new() } }
+    pub(crate) fn lifetime(did: DefId, name: Symbol) -> Self {
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/types.rs:1348]: Same.

**Comment by GuillaumeGomez** [comment]
Looks good to me! Just a nit and it'll be ready to go. :)
